sys/arm64/vmm/vmm_arm64.c

```diff
@@ -0,0 +1,1254 @@
+		el2_vmem_add(next_hyp_va, HYP_VM_MAX_ADDRESS - next_hyp_va);
+
+	daif = intr_disable();
+	cnthctl_el2 = vmm_call_hyp1(HYP_READ_REGISTER, HYP_REG_CNTHCTL);
+	intr_restore(daif);
+
+	vgic_init();
+	vtimer_init(cnthctl_el2);
+
+	return (0);
+}
+
+int
+vmmops_modcleanup(void)
+{
+	int cpu;
+
+	smp_rendezvous(NULL, arm_teardown_vectors, NULL, NULL);
+
+	CPU_FOREACH(cpu) {
+		vmmpmap_remove(stack_hyp_va[cpu], VMM_STACK_PAGES * PAGE_SIZE,
+		    false);
+	}
+
+	vmmpmap_remove(hyp_code_base, hyp_code_len, false);
+
+	vtimer_cleanup();
+
+	vmmpmap_fini();
+
+	CPU_FOREACH(cpu)
+		free(stack[cpu], M_HYP);
+
+	pmap_clean_stage2_tlbi = NULL;
+	pmap_stage2_invalidate_range = NULL;
+	pmap_stage2_invalidate_all = NULL;
+
+	return (0);
+}
+
+static vm_size_t
+el2_hyp_size(struct vm *vm)
+{
+	return (round_page(sizeof(struct hyp) +
+	    sizeof(struct hypctx *) * vm_get_maxcpus(vm)));
+}
+
+static vm_size_t
+el2_hypctx_size(void)
+{
+	return (round_page(sizeof(struct hypctx)));
+}
+
+static vm_pointer_t
+el2_map_enter(vm_offset_t data, vm_size_t size, vm_prot_t prot)
+{
+	vmem_addr_t addr;
+	int err __diagused;
+	bool rv __diagused;
+
+	err = vmem_alloc(el2_mem_alloc, size, M_NEXTFIT | M_WAITOK, &addr);
+	MPASS(err == 0);
+	rv = vmmpmap_enter(addr, size, vtophys(data), prot);
+	MPASS(rv);
+
+	return (addr);
+}
+
+void *
+vmmops_init(struct vm *vm, pmap_t pmap)
+{
+	struct hyp *hyp;
+	vm_size_t size;
+	vm_prot_t prot;
+
+	size = el2_hyp_size(vm);
+	hyp = malloc_aligned(size, PAGE_SIZE, M_HYP, M_WAITOK | M_ZERO);
+
+	hyp->vm = vm;
+	hyp->vgic_attached = false;
+
+	vtimer_vminit(hyp);
+	vgic_vminit(hyp);
+
+	prot = VM_PROT_READ | VM_PROT_WRITE;
+#if __has_feature(capabilities)
+	prot |= VM_PROT_READ_CAP | VM_PROT_WRITE_CAP;
+#endif
+	hyp->el2_addr = el2_map_enter((vm_offset_t)hyp, size, prot);
+
+	return (hyp);
+}
+
+void *
+vmmops_vcpu_init(void *vmi, struct vcpu *vcpu1, int vcpuid)
+{
+	struct hyp *hyp = vmi;
+	struct hypctx *hypctx;
+	vm_size_t size;
+	vm_prot_t prot;
+
+	size = el2_hypctx_size();
+	hypctx = malloc_aligned(size, PAGE_SIZE, M_HYP, M_WAITOK | M_ZERO);
+
+	KASSERT(vcpuid >= 0 && vcpuid < vm_get_maxcpus(hyp->vm),
+	    ("%s: Invalid vcpuid %d", __func__, vcpuid));
+	hyp->ctx[vcpuid] = hypctx;
+
+	hypctx->hyp = hyp;
+	hypctx->vcpu = vcpu1;
+
+	reset_vm_el01_regs(hypctx);
+	reset_vm_el2_regs(hypctx);
+
+	vtimer_cpuinit(hypctx);
+	vgic_cpuinit(hypctx);
+
+	prot = VM_PROT_READ | VM_PROT_WRITE;
+#if __has_feature(capabilities)
+	prot |= VM_PROT_READ_CAP | VM_PROT_WRITE_CAP;
+#endif
+	hypctx->el2_addr = el2_map_enter((vm_offset_t)hypctx, size, prot);
+
+	return (hypctx);
+}
+
+static int
+arm_vmm_pinit(pmap_t pmap)
+{
+
+	pmap_pinit_stage(pmap, PM_STAGE2, vmm_pmap_levels);
+	return (1);
+}
+
+struct vmspace *
+vmmops_vmspace_alloc(vm_offset_t min, vm_offset_t max)
+{
+	vm_pointer_t minp, maxp;
+
+#ifdef __CHERI_PURE_CAPABILITY__
+	minp = (vm_pointer_t)cheri_setaddress(vmm_gpa_root_cap, min);
+	minp = (vm_pointer_t)cheri_setboundsexact(minp, max - min);
```

Comment:
It depends on what you mean by guaranteed. It is today, because the vmspace bounds are fixed and representable ([0, 1<<39]). If someone changes the GPA space constants, this may become false, but there's no reason to use a non-representable range.